### PR TITLE
etcdserver: prune revisions before send backend readTx when rangekeys

### PR DESCRIPTION
**Pull request by xrmzju**
fix [issue 12301](https://github.com/etcd-io/etcd/issues/12301)

### CODE REVIEW

**Comment by xrmzju** [comment]
this improves performance tremendously for request WithLastRev() or WithFristRev() option
without this pr:
```shell
[root@edd2693ca788 /]# time ./etcdctl  get --prefix=true /default/ --last-rev=true --keys-only -wjson --command-timeout=60s
{"header":{"cluster_id":11136150232382961551,"member_id":4898650950304686290,"revision":1337662,"raft_term":23},"kvs":[{"key":"L2RlZmF1bHQvQjFDVFl5d0ZxSVFpb2ZHMXBPcXpiVGtBbWVKTFQyZGc=","create_revision":1337662,"mod_revision":1337662,"version":1}],"more":true,"count":1317250}

real	0m35.135s
user	0m0.031s
sys	0m0.022s
```
with this pr
```


[root@edd2693ca788 /]# time ./etcdctl   get --prefix=true /default/ --last-rev=true --keys-only -wjson
{"header":{"cluster_id":11136150232382961551,"member_id":4898650950304686290,"revision":1337662,"raft_term":22},"kvs":[{"key":"L2RlZmF1bHQvQjFDVFl5d0ZxSVFpb2ZHMXBPcXpiVGtBbWVKTFQyZGc=","create_revision":1337662,"mod_revision":1337662,"version":1}],"more":true,"count":1317250}

real	0m0.892s
user	0m0.027s
sys	0m0.023s
```

**Comment by xrmzju** [comment]
@xiang90 @gyuho @jpbetz @jingyih @spzala please take a look. thx.

**Comment by xrmzju** [comment]
i don't see why the ci build failed ,can somebody help me...

File: mvcc/kvstore_txn.go [hunk 1]
```diff
@@ -15,10 +15,12 @@
...
 import (
-	"go.etcd.io/etcd/v3/lease"
-	"go.etcd.io/etcd/v3/mvcc/backend"
-	"go.etcd.io/etcd/v3/mvcc/mvccpb"
-	"go.etcd.io/etcd/v3/pkg/traceutil"
+	"sort"
+
+	"go.etcd.io/etcd/lease"
+	"go.etcd.io/etcd/mvcc/backend"
+	"go.etcd.io/etcd/mvcc/mvccpb"
```
Comment (cfc4n) [inline mvcc/kvstore_txn.go:22]: Use `go.etcd.io/etcd/v3` replace `go.etcd.io/etcd` please. That is the rootcause of CI failed like https://travis-ci.com/github/etcd-io/etcd/jobs/385763253 .

Comment (xrmzju) [reply mvcc/kvstore_txn.go]: yeah... thank u very much

**Comment by tangcong** [comment]
=== RUN   TestV3RangeRequest
    TestV3RangeRequest: v3_grpc_test.go:1514: #4.3: key[0]. got = a, want = c, 
--- FAIL: TestV3RangeRequest (1.12s)

@xrmzju can you take a look at this failed test case?

**Comment by xrmzju** [comment]
> === RUN TestV3RangeRequest
> TestV3RangeRequest: v3_grpc_test.go:1514: #4.3: key[0]. got = a, want = c,
> --- FAIL: TestV3RangeRequest (1.12s)
> 
> @xrmzju can you take a look at this failed test case?

fixed

**Comment by xrmzju** [comment]
if u guys have any other feedbacks, please let me know, many thanks! @gyuho @jpbetz @jingyih @tangcong @cfc4n

**Pull request closed without merging**